hook-common/src/webhook.rs

```diff
@@ -134,3 +136,99 @@ pub struct WebhookJobMetadata {
     pub plugin_id: Option<i32>,
     pub plugin_config_id: Option<i32>,
 }
+
+/// An error originating during a Webhook Job invocation.
+#[derive(Serialize, Debug)]
+pub struct WebhookJobError {
+    pub r#type: app_metrics::ErrorType,
+    pub details: app_metrics::ErrorDetails,
+    #[serde(serialize_with = "serialize_uuid")]
+    pub uuid: uuid::Uuid,
+}
+
+impl From<&reqwest::Error> for WebhookJobError {
+    fn from(error: &reqwest::Error) -> Self {
+        if error.is_body() || error.is_decode() {
+            WebhookJobError::new_parse(&error.to_string())
+        } else if error.is_timeout() {
+            WebhookJobError::new_timeout(&error.to_string())
+        } else if error.is_status() {
+            WebhookJobError::new_http_status(
+                error.status().expect("status code is defined").into(),
+                &error.to_string(),
+            )
+        } else if error.is_connect()
+            || error.is_builder()
+            || error.is_request()
+            || error.is_redirect()
+        {
+            // Builder errors seem to be related to unable to setup TLS, so I'm bundling them in connection.
+            WebhookJobError::new_connection(&error.to_string())
+        } else {
+            // We can't match on Kind as some types do not have an associated variant in Kind (e.g. Timeout).
```

Comment:
Interesting. I think this this is fine, but could you also check for `is_timeout()` up top and short-circuit, and then match on Kind?

I'm not sure if it would be cleaner or not, up to you.

Reply:
Unfortunately, the `kind` field of the `inner` struct held by `reqwest::Error` is private, so we can't match on it anyways :sweat: 

I'm not entirely sure about what each error kind means so we may need to adjust these as we go along.

Reply:
Damn, it's a shame that a reqwest upgrade could lead to a crash in prod (right? new error type?).

Hopefully that's unlikely. I guess the only other option is to have a catch-all type and then log it loudly somewhere we'd see it.

Reply:
Yeah, a new error type would cause us to panic (assuming such an error is returned somewhere). There is also a risk of bugs in reqwest causing us to panic: Most of the `is_*` methods check for `Kind::{method_name}` to return a `bool`, but some of them do more complicated checks, for example to determine whether an error `is_timeout`. A bug that causes all of the `is_*` methods to return `false` could also cause us to panic.

The catch-all type would work, maybe `Unknown` or something.

Reply:
Alternatively, we could ignore whatever reqwest says the error is and determine the error type based on the call that triggered it, for example:
1. `send()` can only be a `Timeout` or `Connection`.
2. Parsing URL/headers methods can only be a `Parse` error.
3. After receiving a response, error can only be `HttpStatus`.

2 and 3 are fairly straight-forward I think, but 1 is the one with troubles: some error types like `is_builder`, `is_body`, or `is_decode`, may be grouped together with `Connection`.

Reply:
I kind of like that approach more, just to avoid the production panic: I rather have a broad definition of "Connection Error" than crash. Instead of having to add an error type to fix a crash, we can add more error types when needed for more specific reporting.

Reply:
Sounds good to me!